### PR TITLE
Module support

### PR DESCRIPTION
**Pull request by dcb210**
These commits implement module support for Matlab.  Note that I had to introduce a new member variable, swigPtr, to handle cross module pointers.  Also, the mex files must remain locked in order to support the cross-module type table.  A better implementation may exist, but the submitted changes allow me to run our standard test script in Matlab without runtime errors.

### CODE REVIEW

**Comment by jaeandersson** [comment]
Really great you implemented this! Unfortunately, it is breaking for me: I have issues with functions returning SWIG objects. I'll see if I can resolve this: this was something I wanted to refactor anyway.

**Comment by jaeandersson** [comment]
Oh, I see you already refactored that (nice!). Will try again to merge and sort out the bugs.

**Comment by dcb210** [comment]
Yes, the main change was the addition of swigPtr to SwigRef that supplies a pointer to the SWIG_Matlab_GetPointer() function.  The SWIG_Matlab_GetPointer() function was also modified to optionally delete the pointer as well.  I also implemented the SWIG_Matlab _SetModule() & SWIG_Matlab_GetModule() functions.  The alternative would be to use a global variable that stores the pointer to swigPtr similar to SWIG_Matlab _SetModule() & SWIG_Matlab_GetModule().

**Comment by jaeandersson** [comment]
OK... I'm debugging... Found one bug in `SWIG_Matlab_ConvertPtrAndOwn` (should return `SWIG_ERROR` if conversion fails - which is not a fatal error):

```
diff --git a/Lib/matlab/matlabrun.swg b/Lib/matlab/matlabrun.swg
index 9442706..58534c0 100644
--- a/Lib/matlab/matlabrun.swg
+++ b/Lib/matlab/matlabrun.swg
@@ -346,10 +346,7 @@ SWIGRUNTIME int SWIG_Matlab_isOwned(mxArray* pm) {
 SWIGRUNTIME int SWIG_Matlab_ConvertPtrAndOwn(mxArray* pm, void **ptr, swig_type_info *type, int flags, int *own) {
   /* Get pointer object */
   SwigPtr* p = SWIG_Matlab_getSwigPtr(pm);
-  if (!p) {
-    mexWarnMsgIdAndTxt("SWIG:ConvertPtrAndOwn", "No such element");
-    return 0;
-  }
+  if (!p) return SWIG_ERROR;

   swig_cast_info *tc;
   bool same_type = false;
```

It gets me further, but I still have issues.

**Comment by jaeandersson** [comment]
The second issue has to do with `_swigCreate` it seems, which should be refactored away anyway.

**Comment by dcb210** [comment]
I didn’t see a good way to do that.  Since it has to do with ownership and class inheritance.  For inheritance we don’t want the Matlab base constructor to create a C++ instance of the base object.  Also internally we create Matlab objects that are references to C++ objects that are not owned by Matlab.  In those instances we don’t want to create a duplicate C++ object.

**Comment by dcb210** [comment]
Let me know if I can help in any way.
-Don

**Comment by jaeandersson** [comment]
I'm trying to make a proper fix of this. My approach is that the "newobject" helper class returns a SwigRef instance. This is definitely cleaner and more efficient. Only thing left to resolve is functions returning class instances. Here the proper approach is to use MATLAB's "No Input Argument Constructor Requirement" #31. I'm convinced that this is the correct approach both for efficiency and style. I'll get back when it's working properly.

**Comment by jaeandersson** [comment]
Note that this is something that should be done regardless of this issue.

File: Lib/matlab/matlabrun.swg [hunk 1]
```diff
@@ -204,65 +270,72 @@ void SWIG_Matlab_ExitFcn(void) {
...
   }
 
-  /* Pass to class */
-  mxArray *pm_ind = mxCreateNumericMatrix(1, 1, mxUINT64_CLASS, mxREAL);
-  *(uint64_T *)mxGetData(pm_ind) = (uint64_T)ind;
-  if(!pm_ind) mexErrMsgIdAndTxt("SWIG:NewPointerObj","mxCreateNumericMatrix failed");
-  mxSetProperty(pm, 0, "swigInd", pm_ind);
-  mxDestroyArray(pm_ind);
+  /* Different behavior depending on ownership */
+  if (!own) {
```
Comment (jaeandersson) [inline Lib/matlab/matlabrun.swg:285]: @dcb210 This design won't work. It is breaking for all functions that return object instances:

``` c++
MyWrappedClass myfunction();
```

Comment (dcb210) [inline Lib/matlab/matlabrun.swg:285]: Yes, I’m thinking that we should use a global variable to store swigPtr.  The first module to load then manages storage for all other modules.  I can make changes locally in my fork and send a pull request.

**Comment by jaeandersson** [comment]
I think a better solution is to go back to have "swigInd" be a pointer (instead of just a reference). I think you proposed that earlier. That pointer would then in turn contain a pointer to the module. That would be more efficient also since less communication goes via MATLAB.

I just noticed a serious memory leak (basically, the destructors don't get called in many cases). Looking at the codegen, I noticed that in the wrapped destructor, the ownership flag is first removed and then a check is done whether to delete the object... Which is never done of course... I will try to resolve this first. It should be pretty straightforward.

**Comment by jaeandersson** [comment]
Memory leak (269a40858f5eb218c67c0b15b8e6b5e1afdb4319) resolved. I'll change back the swigInd type next. Hopefully it's possible without getting back the segfaults...

**Comment by jaeandersson** [comment]
I replaced swigInd with a raw pointer and added a `swig_module_info *` as a member of the SwigRef struct (3586dbf18c4d407b8321e0dafd24335002e28dc1). This should take care of the storage issue.

If you want to take another shot at the module, free free. Otherwise I can try to do it in the next couple of weeks (this week a bit too busy).

**Comment by dcb210** [comment]
Actually, adding swig_module_info to SwigRef won’t solve the storage issue.  See SWIG_InitializeModule() and swig/Doc/Devel/runtime.txt.  SWIG is designed so that all of the type information should be global.  When SWIG_InitializeModule() is called it adds the type information to the current global type information which, by default, is stored in the global swig_module variable of the first module to load.  The current design approach in Matlab is to unload the mex when all of objects it creates are deleted.  This will destroy the swig_module variable and will lead to crashes when other modules attempt to access the global type information.  Without creating a completely different function to replace SWIG_InitializeModule(), something I believe should be avoided, the Matlab implementation should change so that all modules (and by extension mex files) are loaded once and locked permanently for the currently running Matlab process.  That’s what I implemented in the pull request.  Do you have a better approach in mind that doesn’t require reimplementing SWIG supplied functions?

**Comment by jaeandersson** [comment]
I see... I think I need to study closer how it actually works. There is a script that is run at initialization of a model. One could add to it a "registration" of the module to a MATLAB "persistent" variable, located in the SwigRef base class. I'll have another look and get back.

In any case, do you have a solution to functions returning (owning) swig references? Did you try to run the test suite on your branch?

**Comment by dcb210** [comment]
Correct, I modified the *setup.m script to load all dependent modules.  I did not run the test suite.  How do I run the test suite?

**Comment by dcb210** [comment]
I’m trying ‘make check’ now.

**Comment by jaeandersson** [comment]
See #6: 

```
> make  check-matlab-test-suite RUNPIPE=\>\>/tmp/xxx
> grep failed /tmp/xxx | wc -l
> grep passed /tmp/xxx | wc -l
```

**Pull request closed without merging**